File: ios/RNOkaySdk.m

#import "RNOkaySdk.h"

#import <PSA/PSA.h>

@implementation RNOkaySdk

RCT_EXPORT_MODULE();

RCT_REMAP_METHOD(authorization,
                 spaAuthorizationData: (NSDictionary *) spaAuthorizationData
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    @try{
        NSNumber *sessionId = [spaAuthorizationData valueForKeyPath:@"SpaAuthorizationData.sessionId"];
        [PSA startAuthorizationWithTheme: nil
                               sessionId: sessionId
                              completion:^(BOOL result, PSASharedStatuses status) {
                                  NSString *response = [[NSString alloc] init];
                                  if(status != Success) {
                                      response = @"Authorization finished: Error";
                                  } else {
                                      response = @"Authorization finished: OK";
                                  }
                                  resolve(response);
                              }];
    }
    @catch (NSError *error) {
        reject(@"Error",@"Error", error);
    }
}


RCT_REMAP_METHOD(enrollProcedure,
                 spaEnrollData: (NSDictionary *)spaEnrollData
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    @try{
        NSString *appPns = [spaEnrollData valueForKeyPath:@"SpaEnrollData.appPns"];
        NSString *host = [spaEnrollData valueForKeyPath:@"SpaEnrollData.host"];
        NSString *installationId = [spaEnrollData valueForKeyPath:@"SpaEnrollData.installationId"];
        NSString *pubPss = [spaEnrollData valueForKeyPath:@"SpaEnrollData.pubPss"];
        [PSA startEnrollmentWithHost: host
                      installationId: installationId
                        pubPssBase64: pubPss
                        idCompletion:^(PSASharedStatuses status) {
                            NSString *response = [[NSString alloc] init];
                            if(status != Success) {
                                response = @"Enrollment finished: Error";
                            } else {
                                response = @"Enrollment finished: OK";
                            }
                            resolve(response);
                        }];
    }
    @catch (NSError *error) {
        reject(@"Error",@"Error", error);
    }
}


RCT_REMAP_METHOD(isEnrolled,
                 isEnrolledWithResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    @try{
        BOOL isEnrolled = [PSA isEnrolled];
        resolve(isEnrolled ? @"true" : @"false");
    }
    @catch (NSError *error) {
        reject(@"Error",@"Error", error);
    }
}


RCT_REMAP_METHOD(updateDeviceToken,
                 token: (NSString *) token
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    @try{
        [PSA updateDeviceToken:token];
        
    }
    @catch (NSError *error) {
        reject(@"Error",@"Error", error);
    }
}

RCT_REMAP_METHOD(isReadyForAuthorization,
                 isReadyForAuthorizationdWithResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
    @try{
        BOOL isReadyForAuthorization = [PSA isReadyForAuthorization];
        resolve(isReadyForAuthorization ? @"true" : @"false");
    }
    @catch (NSError *error) {
        reject(@"Error",@"Error", error);
    }
}

@end

File: ios/RNOkaySdk.swift
//
//  RNOkaySdk.swift
//  RNOkaySdk
//
//  Created by Meer, Pavel on 4/18/19.
//  Copyright © 2019 Facebook. All rights reserved.
//

import Foundation
//import PSA
//import PSACommon
//
@objc(RNOkaySdk)
class RNOkaySdk: NSObject {
//
//        @objc func updateDeviceToken(deviceToken: String!, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
//            do {
//                PSA.updateDeviceToken(deviceToken)
//            } catch {
//                reject("Failed update device token", "Failed update device token", error);
//            }
//
//            resolve("Succes update device token")
//        }
//
//        @objc func enrollmentProcedure(withHost: String, installationId: String, pubPssBase64: String, resolve: @escaping RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
//            do {
//                if PSA.isReadyForEnrollment() {
//                    try PSA.startEnrollment(withHost: withHost, installationId: installationId, pubPssBase64: pubPssBase64) { status in
//                        resolve(status)
//                    }
//                }
//            } catch {
//                reject("setup_audio_session_failed", "Failed to setup audio session", error)
//            }
//        }
//
//        @objc func isEnrolled( _ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
//            do {
//                let isEnrolled : Bool = PSA.isEnrolled();
//                resolve(isEnrolled);
//            } catch {
//                reject("setup_audio_session_failed", "Failed to setup audio session", error)
//            }
//        }
//
//    //    @objc func isEnrolled() -> Bool {
//    //       return PSA.isEnrolled();
//    //    }
//
//        @objc func isReadyForAuthorization(resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
//            do {
//                let isReadyForAuthorization : Bool = PSA.isReadyForAuthorization();
//                resolve(isReadyForAuthorization);
//            } catch {
//                reject("setup_audio_session_failed", "Failed to setup audio session", error)
//            }
//        }
//
////        @objc(authorization:sessionId:resolver:rejecter:)
////        func authorization(with: PSATheme?, sessionId:NSNumber?, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
////            do {
////                PSA.startAuthorization(with: PSATheme, sessionId:NSNumber?) { isCancelled, status in
////                    resolve(status)
////                }
////            } catch {
////                reject("setup_audio_session_failed", "Failed to setup audio session", error)
////            }
//    //        }
}
